Plugins/Gradle/src/main/kotlin/com/shopify/testify/TestifyExtension.kt

```diff
@@ -59,61 +59,65 @@ internal data class TestifySettings(
-            this.testPackageId = extension.testPackageId!!
-        }
-        if (extension.applicationPackageId?.isNotEmpty() == true) {
-            this.targetPackageId = extension.applicationPackageId!!
-        }
-        if (extension.installTask?.isNotEmpty() == true) {
-            this.installTask = extension.installTask
-        }
-        if (extension.installAndroidTestTask?.isNotEmpty() == true) {
-            this.installAndroidTestTask = extension.installAndroidTestTask
-        }
-        if (extension.autoImplementLibrary != null) {
-            this.autoImplementLibrary = extension.autoImplementLibrary!!
+    internal companion object {
+        fun create(project: Project): TestifySettings {
+            val android = project.android
+            val extension = project.getTestifyExtension()
+
+            val assetsSet = android.sourceSets.getByName("""androidTest""").assets
+            val baselineSourceDir = extension.baselineSourceDir ?: assetsSet.srcDirs.first().path
+            val testRunner = extension.testRunner ?: android.defaultConfig.testInstrumentationRunner ?: "unknown"
+            val pullWaitTime = extension.pullWaitTime ?: 0L
+            val testPackageId = extension.testPackageId ?: project.inferredDefaultTestVariantId
+            val targetPackageId = extension.applicationPackageId ?: project.inferredTargetPackageId
+            val version = TestifySettings::class.java.getPackage().implementationVersion
+            val autoImplementLibrary = extension.autoImplementLibrary
+                ?: version?.contains("local", ignoreCase = true) == false
+            val useSdCard = extension.useSdCard ?: false
+            val installTask = extension.installTask ?: project.inferredInstallTask
+            val outputFileNameFormat = extension.outputFileNameFormat
+            val installAndroidTestTask = extension.installAndroidTestTask ?: project.inferredAndroidTestInstallTask
+            val moduleName = extension.moduleName ?: project.name
+
+            return TestifySettings(
+                baselineSourceDir = baselineSourceDir,
+                moduleName = moduleName,
+                outputFileNameFormat = outputFileNameFormat,
+                pullWaitTime = pullWaitTime,
+                testRunner = testRunner,
+                useSdCard = useSdCard,
+                testPackageId = testPackageId,
+                targetPackageId = targetPackageId,
+                installTask = installTask,
+                installAndroidTestTask = installAndroidTestTask,
+                autoImplementLibrary = autoImplementLibrary
+            )
         }
-        return this
     }
 
     fun validate() {
-        if (targetPackageId.isEmpty()) {
-            throw GradleExtensionException(
-                """
+        val propertyName = when {
+            targetPackageId.isEmpty() -> "applicationPackageId"
+            testPackageId.isEmpty() -> "testPackageId"
+            else -> null
+        }
```

Comment:
How about something like this?

```suggestion
        val (propertyName, example) = when {
            targetPackageId.isEmpty() -> "applicationPackageId" to "com.example.app"
            testPackageId.isEmpty() -> "testPackageId" to "com.example.app.test"
            else -> null to null
        }
```

Then we can use `$it "$example" down on L117 below